src/rgw/rgw_reshard.cc

```diff
@@ -748,12 +871,57 @@ int RGWReshardWait::block_while_resharding(RGWRados::BucketShard *bs, string *ne
       return 0;
     }
     ldout(store->ctx(), 20) << "NOTICE: reshard still in progress; " << (i < num_retries - 1 ? "retrying" : "too many retries") << dendl;
-    /* needed to unlock as clear resharding uses the same lock */
 
     if (i == num_retries - 1) {
       break;
     }
 
+    // if bucket is erroneiously marked as resharding (e.g., crash or
+    // other error) then fix it. If we can take the bucket reshard
+    // lock, then it means no other resharding should be taking place,
+    // and we're free to clear the flags.
+    {
+      // since we expect to do this rarely, we'll do our work in a
+      // block and erase our work after each try
+
+      RGWObjectCtx obj_ctx(bs->store);
+      const rgw_bucket& b = bs->bucket;
+      std::string bucket_id = b.get_key();
+      RGWBucketInfo bucket_info;
+      ret = bs->store->get_bucket_info(obj_ctx, b.tenant, b.name,
+				       bucket_info, nullptr, nullptr);
```

Comment:
RGWRados::Bucket::UpdateIndex already has access to the bucket info through target->get_bucket_info(), and UpdateIndex::guard_reshard() should be able to pass that in as an argument to block_while_resharding()

Reply:
@cbodley The above comment has my audit. So I allowed the BucketShard::init function to pass out the RGWBucketInfo object it creates so it could be captured by the caller. With that RGWIndexCompletionThread::process can get an RGWBucketInfo and now we can pass that object into RGWRados::guard_reshard and to RGWRados::block_while_resharding and finally to RGWReshardWait::block_while_resharing.